Make JSON printing a bit nicer

gdb opened the pull request:
This change makes JSON generation more consistent and human-friendly:

- Adding a trailing newline is nicer to people using `curl` from the command line, without making the machine case worse.
- Add `indent=2` to errors so they are pretty printed as well.

jmcs commented on connexion/problem.py:
```diff
@@ -46,7 +46,7 @@ def problem(status, title, detail, type='about:blank', instance=None, headers=No
         problem_response.update(ext)
 
-    body = json.dumps(problem_response)
+    body = json.dumps(problem_response, indent=2) + "\n"
```
The `+ "\n"` means you create 3 string every time instead of 1.

jmcs commented:
The unit tests are also failing.

gdb commented:
Whoops, sorry yes I'd meant to ask about whether the implementation and functionality was ok before fixing up the tests. (On the one hand, copies always sound bad. On the other hand, there was just a `json.dumps` there, which I'd expect would always be much more expensive than the copy. You also mention 3 strings — are you concerned about creating the '\n'?)

I just switched the implementation to not do the copy (but instead pass a list to the Response — I'm less familiar with the Python WSGI ecosystem than the Ruby Rack one, but I believe this should be supported regardless of `response_class`?), and fixed the tests — is this any better?

ainmosni commented:
Is it really the frameworks responsibility to pretty print it? I mean, everyone working with REST has tools to pretty print it on their end. At the very least they can pipe it to python -m json.tool or jq.

gdb commented:
At Stripe, one of our core observations was that the little details are what changes an API from acceptable to actually a pleasure to use. (And particularly, the most important details happen during interactive development: when you're first `curl`ing around a new API. We would do things like pre-fill your test API key in the API docs, so you can just copy-paste the examples directly without having to figure out where your API key lives, etc.)

Certainly developers can pipe the output of a curl to `jq`. But even better would be anticipating this use-case and returning the JSON already in a human-friendly format.

Regardless of the philosophy, this framework already does `indent=2` in one case; here I've made it consistent and added trailing newlines. I'd also be equally happy if this functionality were easily pluggable. But I think it's actually a pretty good thing for the framework to provide little finishing touches that otherwise most people probably wouldn't bother with.

hjacobs commented:
@gdb maybe you can consider using HTTPie, https://github.com/jkbrzt/httpie

>  .. user-friendly curl replacement with intuitive UI, JSON support, syntax highlighting, wget-like downloads, extensions, etc

I use it everyday to do REST calls :smile:

gdb commented:
HTTPie is great. Unfortunately, it's less about me developing my own API, and more about the external developers who are going to integrate against my API.

hjacobs commented:
@jmcs @rafaelcaricio what is now the status here (apart from merge conflicts)? I think the trailing newline is OK (we already have indent as @gdb points out).

rafaelcaricio commented:
@hjacobs I guess it is fine to add the new-line. In the future we could add configuration to Connexion to enable "human readable JSON responses", so it would enable the indentation + newline.

hjacobs commented:
@gdb can you resolve the merge conflict to get this PR into master? Thanks!

gdb commented:
Ok!
